src/main/scala/latis/service/hapi/DataService.scala

```diff
@@ -25,33 +26,44 @@ class DataService[F[_]: Concurrent](
   val service: HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "data"
-          :? IdMatcher(_id)
+          :? DatasetMatcher(_dataset)
+          +& IdMatcher(_id)
+          +& StartTimeMatcher(_startTime)
+          +& StopTimeMatcher(_stopTime)
           +& MinTimeMatcher(_minTime)
           +& MaxTimeMatcher(_maxTime)
           +& ParamMatcher(_params)
           +& IncludeMatcher(_inc)
           +& FormatMatcher(_fmt) =>
-        val req: Either[Status, DataRequest] =
+        val req: Either[Status, DataRequest] = {
+          // For backwards compatibility, allow both
+          // "dataset" or "id",
+          // "start" or "time.min",
+          // "stop" or "time.max"
           for {
-            id      <- _id.asRight
-            minTime <- _minTime.leftMap(_ => Status.`1402`).toEither
-            maxTime <- _maxTime.leftMap(_ => Status.`1403`).toEither
-            _       <- Either.cond(minTime.isBefore(maxTime), (), Status.`1404`)
-            params  <- Either.cond(
+            dataset   <- Either.fromOption(_dataset <+> _id, Status.`1400`)
+            startTime <- Either.fromOption(_startTime <+> _minTime, Status.`1400`)
+              .flatMap(_.leftMap(_ => Status.`1402`).toEither)
+            stopTime  <- Either.fromOption(_stopTime <+> _maxTime, Status.`1400`)
+              .flatMap(_.leftMap(_ => Status.`1403`).toEither)
+            _         <- Either.cond(startTime.isBefore(stopTime), (), Status.`1404`)
+            params    <- Either.cond(
               _params.map(_.count(_ != "time") > 0).getOrElse(true),
               _params,
               Status.`1501`
             )
-            inc     <- _inc.getOrElse(Include(false).validNel).bimap(
+            inc       <- _inc.getOrElse(Include(false).validNel).bimap(
               _ => Status.`1410`, _.header
             ).toEither
-            fmt     <- _fmt.getOrElse(Format("csv").validNel).bimap(
+            fmt       <- _fmt.getOrElse(Format("csv").validNel).bimap(
               _ => Status.`1409`, _.format
             ).toEither
-          } yield DataRequest(id, minTime, maxTime, params, inc, fmt)
+          } yield DataRequest(dataset, startTime, stopTime, params, inc, fmt)
+        }
         val records: EitherT[F, Status, Stream[F, String]] = for {
-          req     <- EitherT.fromEither[F](req)
-          header  <- alg.getMetadata(_id, _params).leftMap {
+          req    <- EitherT.fromEither[F](req)
+          dataset = _dataset.getOrElse(_id.getOrElse(""))
```

Comment:
Now that how we get the dataset identifier has changed, I think it would be cleaner to get the dataset ID from `req`.

Reply:
Oh, agreed.